jmespath-core/src/main/java/io/burt/jmespath/node/JsonLiteralNode.java

```diff
@@ -3,35 +3,33 @@
 import io.burt.jmespath.Adapter;
 
 public class JsonLiteralNode<T> extends Node<T> {
-  private final String raw;
-  private final T tree;
+  private final String rawJson;
+  private final T json;
 
-  public JsonLiteralNode(Adapter<T> runtime, String raw, T tree) {
+  public JsonLiteralNode(Adapter<T> runtime, String rawJson) {
     super(runtime);
-    this.raw = raw;
-    this.tree = tree;
+    this.rawJson = rawJson;
+    this.json = runtime.parseString(rawJson);
   }
 
   @Override
   public T search(T input) {
-    return tree;
+    return json;
   }
 
   @Override
   protected String internalToString() {
-    return String.format("%s", raw);
+    return String.format("%s", rawJson);
   }
 
   @Override
   protected boolean internalEquals(Object o) {
     JsonLiteralNode<?> other = (JsonLiteralNode<?>) o;
-    return tree.equals(other.tree);
+    return json.equals(other.json);
```

Comment:
I personally think that going with either extreme is the right choice. Either making `#equals`, `#hashCode`, and `#toString` work with the raw input (in both classes), or using the parsed value, but then I think equality should use `runtime.compare(x, other.x) == 0` construct for `#equals`, `runtime.toString(x)` for `#toString`, and we would probably need a new construct to implement `#hashCode` properly. I don't think it is a good idea to rely on `T#equals`, `T#hashCode`, and `T#toString` to represent JSON semantics.